lib/spack/spack/monitor.py

```diff
@@ -470,13 +470,13 @@ def get_build_id(self, spec, return_response=False, spec_exists=True):
         """
         Retrieve a build id, either in the local cache, or query the server.
         """
-        full_hash = spec.full_hash()
-        if full_hash in self.build_ids:
-            return self.build_ids[full_hash]
+        dag_hash = spec.dag_hash()
+        if dag_hash in self.build_ids:
+            return self.build_ids[dag_hash]
 
         # Prepare build environment data (including spack version)
         data = self.build_environment.copy()
-        data['full_hash'] = full_hash
+        data['hash'] = dag_hash
```

Comment:
I think we should just do this:

```suggestion
        data['full_hash'] = dag_hash
```

So that we keep the same keys that monitor currently understands.  We can update this later, but this at least will keep the protocol with `spack monitor` the same.

I think the rest of the changes in this file are fine, because the `dag_hash` now *is* the `full_hash`.

@vsoch do you agree with this and my comment in the tests?

Reply:
Yes and also 1. undo the changes to the mock, and 2. check the spec objects at each endpoint request and ensure there is still a full_hash. It's fine if it's a dag hash spack monitor just needs that field and treats it as a main identifier.

Reply:
@vsoch: Finally getting back to this -- I just pushed what I think are the needed reverts (see 3fe26353).  Can you take a look?

The part I am not sure I got right is (2).  When you say "spec objects" do you mean the JSON payload to monitor, or do you mean that `Spec` needs to still have a `full_hash` attribute?  If the latter then I probably need to make more changes, but I'd rather get rid of `full_hash` on `Spec` for good if I can.

Reply:
You can basically assume that any previous values (before these changes) should remain the same. This means:

- the build environment is expecting [full_hash](https://github.com/spack/spack-monitor/blob/89acf94dc664b598d9f73292ae4d61fdccf3ac5b/spackmon/apps/api/views/builds.py#L46)
- dependencies need `full_hash` or `build_hash` (and if you change the key it's just going to mess up relationships): https://github.com/spack/spack-monitor/blob/89acf94dc664b598d9f73292ae4d61fdccf3ac5b/spackmon/apps/main/tasks.py#L187
- specs are expected to have [full_hash](https://github.com/spack/spack-monitor/blob/89acf94dc664b598d9f73292ae4d61fdccf3ac5b/spackmon/apps/main/tasks.py#L201).

You can look in that tasks.py file to see all of the import logic. It's probably fine if you want to break things because (afaik) nobody is using spack monitor, and it's use case for research is mostly over (we should take down both servers imho).

Reply:
> Spec needs to still have a full_hash attribute

That one.

Reply:
ok this at the very least is going to break:
> dependencies need `full_hash` or `build_hash` (and if you change the key it's just going to mess up relationships)

Dependencies are just keyed by `"hash"`.

We could *write* the spec JSON sent to spack monitor differently if that would fix this... can you point me at the part of monitor that uses `Spec`?

Reply:
it's that entrypoint that parses the json I linked to - so you want:

- [the builds endpoints](https://github.com/spack/spack-monitor/blob/main/spackmon/apps/api/views/builds.py)
- [the tasks they use](https://github.com/spack/spack-monitor/blob/main/spackmon/apps/main/tasks.py)

Reply:
@tgamblin I think it would be good to get a sense of (if anyone) is using either spack monitor server (I am not, and autamus is going to be refactored I believe and it can be removed there) and we should take them both down. Harshitha and I are working on models that could justify a version III, and if they are successful / look promising we will try for the third round. And that way you can do the changes here that you think are best, and I'll debug in detail and make sure it works for the next round. Thoughts?

Reply:
I think that's a good idea, but I also think I can keep it working here, so I don't think we need to do it immediately.

>> Spec needs to still have a full_hash attribute
>
> That one.

I looked at the two places you linked, and it looks like they do not actually use `spack.spec.Spec` (which is what I was getting at and was afraid of).  They use the JSON we output, so they're fairly loosely coupled with this refactor.  I can make spack monitor send the right JSON and still get rid of `full_hash` on `Spec` in `spack`.

The fix on the monitor side for this would be to make it accept plain old `"hash"` (for spec nodes and dependency keys) in addition to `"full_hash"` or `"build_hash"`, but it's better to let that happen later (if at all).

Reply:
Ok I pushed a change that should preserve the JSON as it was before -- or at least it will keep the `"full_hash"` key for both the nodes and the dependencies.

Reply:
> and it looks like they do not actually use spack.spec.Spec 

yep that's correct - spack is not a dependency of spack monitor (it's used in an external example script but nowhere on the server). The spack spec model is just represented in the database design. Adding spack as a dependency is generally really challenging.